chapter4.adoc

```diff
@@ -221,9 +224,81 @@ follows:
+an access-fault exception corresponding to the original access type.
+
+3. If any bits or encodings that are reserved for future standard use are
+set within _mpte_, stop and raise an access-fault exception corresponding to
+the original access type.
+
+4. Otherwise, the _mpte_ is valid. If (_i_=0) or (_i_=1 and _mpte.type_ is not
+`MTT_L1_DIR`), go to step 5. Otherwise, the _mpte_ is a pointer to the next
+level of the `MTT`. Let _i_ = _i_-1. If _i_ < 0, stop and raise an access-fault
+exception corresponding to the original access type. Otherwise, let
+_a_ = _mpte.ppn_ x _PAGESIZE_ and go to step 2. Note that when _mpte.type_ =
+`MTT_L1_DIR`, the _mpte.ppn_ field is the value of the _mpte.info_ field.
+
+5. A leaf _mpte_ has been found. If any bits or encodings within _mpte.type_
+and _mpte.info_ that are reserved for future standard use, per
+<<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>, are set within _mpte_,
+stop and raise an access-fault exception corresponding to the access type.
+
+6. The _mpte_ is a valid leaf _mpte_. Fetch the access-permissions for the
+physical address per the steps described below:
+
+* if _i_=1, and the _mpte.type_ field specifies it holds the access-permission
+encoding for 1GB pages per <<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>
+; go to step 7, else
+
+* if _i_=1, and the _mpte.type_ field specifies it holds the access-permission
+encoding for 2MB pages; the _mpte.info_ field contains the access-permission
+encoding per <<Smmtt-rw-l2-encoding>> and <<Smmtt-ad-l2-encoding>>; go to
+step 7, else
+
+* If _i_=0, the _mpte_ contains a 2-bit entry for mode `Smmtt[34 | 46 | 56]`
+that holds the access-permission encodings (or 4-bit entry for mode
+`Smmtt[34 | 46 | 56]rw`) for 4KB pages. The 2/4-bit access-permission encoding
+for the PA is held in the ppn referenced by _mpte.info_ and indexed via
+_pa.pn[i]_. The encodings are specified in <<Smmtt-rw>> and <<Smmtt-ad>> for the
+`Smmtt[34 | 46 | 56]` and `Smmtt[34 | 46 | 56]rw` modes respectively.
+
+7. Determine if the requested physical memory access is allowed per the
+access-permissions. If access is not permitted, stop and raise a page-fault
+exception corresponding to the original access type. For modes
+`Smmtt[34 | 46 | 56]`, access-allowed implies read, write and execute
+permissions. For modes `Smmtt[34 | 46 | 56]rw`, read access implies read and
+execute, and read-write permission implies read and write (but no execute).
+
+8. The access is allowed per the `MTT` lookup.
+
+All implicit accesses to the non-leaf memory tracking table data structures in
+this algorithm are performed using width _NON_LEAF_PTE_SIZE_.
+
+[NOTE]
+====
+Implementations may cache any translation (due to 1st stage and/or G-stage
+translation) and access-permissions for the final (leaf) physical address
+being accessed - in this case, the `MTT`-derived access-permissions are
+AND'ed to accumulate the effective read, write, and execute permissions. Thus,
```

Comment:
Smmtt should not prescribe microarchitecture. ANDing with effective read/write/execute permissions will be wrong - consider single stage page table for simplicity. If the TLB has a r, w, and x bit - and say the single stage PTE grants them all but the MTT only grants read/execute permission. With anding it would lead to TLB holding R=1, W=0, X=1. Now if a store hits that TLB entry should implementation raise a page-fault or should it raise an access fault?

Reply:
reduced the note to the main part:

[NOTE]
====
MTT access-permissions can only further restrict access, and never grant
read, write or execute permission denied by 1st-stage or G-stage translations.
====